util.py

```diff
@@ -301,3 +303,73 @@ def instanciate_delayed(o):
         for k in o:
             o[k] = instanciate_delayed(o[k])
     return o
+
+
+already_printed_gs_warnings = set()
+
+
+def get_executable_path(
+    path: Optional[tk.Path],
+    gs_member_name: str,
+    default_exec_path: Optional[tk.Path] = None,
+) -> tk.Path:
+    """
+    Helper function that allows to select a specific version of software while
+    maintaining compatibility to different methods that were used in the past to select
+    software versions.
+    It will return a Path object for the first path found in
+
+    :param path: Directly specify the path to be used
+    :param gs_member_name: get path from sisyphus.global_settings.<gs_member_name>
+    :param default_exec_path: general fallback if no specific version is given
+    """
+    global already_printed_gs_warnings
+    if path is not None:
+        if isinstance(path, tk.Path):
+            return path
+        elif isinstance(path, str):
+            logging.warning(
+                f"use of str is deprecated, please provide a Path object for {path}"
+            )
+            return tk.Path(path)
+        assert False, f"unsupported type of {type(path)} for input {path}"
+    if getattr(gs, gs_member_name, None) is not None:
+        if gs_member_name not in already_printed_gs_warnings:
+            logging.warning(
+                f"use of gs is deprecated, please provide a Path object for gs.{gs_member_name}"
+            )
+            already_printed_gs_warnings.add(gs_member_name)
+
+        return tk.Path(getattr(gs, gs_member_name))
+    if default_exec_path is not None:
+        return default_exec_path
+    assert False, f"could not find executable for {gs_member_name}"
+
+
+def get_returnn_root(returnn_root: tk.Path) -> tk.Path:
+    """gets the path to the root folder of RETURNN"""
+    return get_executable_path(returnn_root, "RETURNN_ROOT")
```

Comment:
Should we provide a `default_returnn_root` if nothing is given? e.g.
```suggestion
    default_root = CloneGitRepositoryJob("https://github.com/rwth-i6/returnn").out_repository
    return get_executable_path(returnn_root, "RETURNN_ROOT", default_root)
```
I don't like it because:
  * the user looses the control over the version that is checked out
  * updating to a more recent version becomes painful / manual work
  * other users that rerun the pipeline might get different results because their HEAD differs

One could instead hardcode the current HEAD, but this might not be very future proof.

Reply:
I think we should have no default, it is better if this is slightly more verbose. I (and I think Albert did something similar independently) have now a `default_tools.py` file, which has content like:

```python
RETURNN_ROOT = CloneGitRepositoryJob("https://github.com/rwth-i6/returnn", commit="45fad83c785a45fa4abfeebfed2e731dd96f960c").out_repository
RETURNN_ROOT.hash_overwrite = "LIBRISPEECH_DEFAULT_RETURNN_ROOT"
```
So yes there is additional overhead in passing this to all the jobs, but it offers:
 - updating without hash change
 - updating with hash change
 - running specific experiments with a different version but same hash
 - running specific experiments with a a different version and different hash